### PR TITLE
feat: add random condition loop wrapper

### PR DESCRIPTION
**Pull request by dechlorine**

### CODE REVIEW

File: src/prog_questions/QuestionRandomCondition.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+"""
+
+class QuestionRandomCondition(QuestionBase):
+
+    def __init__(self, *, seed: int, condition_length: int, array_length: int):
+        super().__init__(seed=seed, condition_length=condition_length, array_length=array_length)
+        self.task = Task(array_length, condition_length, seed)
+        self.sample_code = self.task.code
+        self.parse(self.task.text)
+        random.seed(seed)
```
Comment (R3dKar) [inline src/prog_questions/QuestionRandomCondition.py:80]: `random.seed` лучше вызывать в методе `test`. В нашем случае оно будет итак работать, так как на moodle после конструктора метод `test` вызывается ровно один раз

**Review by R3dKar — APPROVED** [review]

File: src/prog_questions/QuestionRandomCondition.py (hunk 1 above)
Comment (dechlorine) [reply src/prog_questions/QuestionRandomCondition.py:80]: Реализую в рамках решения ишшью "Настройка параметров въедливости и сложности в классе QuestionRandomCondition" (#40)

File: src/prog_questions/QuestionRandomCondition.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        printf("%lld ", arr[i]);
+    }}
+
+    return 0;
+}}
+"""
+
+class QuestionRandomCondition(QuestionBase):
+
+    def __init__(self, *, seed: int, condition_length: int, array_length: int, strictness: float):
```
Comment (NaGIBIN-IGOR) [inline src/prog_questions/QuestionRandomCondition.py:75]: Желательно добавить описание параметров:
"""
    :param seed: Seed для воспроизводимости тестов.
    :param condition_length: Длина условия задачи.
    :param array_length: Длина массива данных.
    :param strictness: Параметр для регулирования количества случайных тестов (0.0 - минимум, 1.0 - максимум).
    """

File: src/prog_questions/QuestionRandomCondition.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                return "OK"
+            else:
+                return f"Ошибка: ожидалось '{expected_output}', получено '{output}'"
+        except CompilationError as e:
+            return f"Ошибка компиляции: {e}"
+        except ExecutionError as e:
+            return f"Ошибка выполнения (код {e.exit_code}): {e}"
+
+    # form test: INT edge case
+    def test_int_edge_case(self, code: str) -> str:
```
Comment (NaGIBIN-IGOR) [inline src/prog_questions/QuestionRandomCondition.py:141]: Рассмотреть больше краевых случаев, к примеру:
- Пустой массив.
- Массив с одним элементом.
- Массив с одинаковыми элементами.
- Массив с чередующимися большими и маленькими числами.

File: src/prog_questions/QuestionRandomCondition.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+    # test
+    def test(self, code: str) -> str:
+        random.seed(self.seed)
+
+        test_result_int_edge = self.test_int_edge_case(code)
+        if test_result_int_edge != "OK":
+            return test_result_int_edge
+
+        random_test_borders = [2, 3, 4, 5, 7, 9]
```
Comment (NaGIBIN-IGOR) [inline src/prog_questions/QuestionRandomCondition.py:167]: Можно создать отдельный метод distribute_tests, который будет заниматься только распределением тестов между границами. Этот метод будет принимать общее количество тестов (total_tests) и список границ (borders) и возвращать массив с количеством тестов для каждой границы.
Пример реализации:
```
def distribute_tests(self, total_tests: int, borders: list) -> list:
    base_count = total_tests // len(borders)  # Базовое количество тестов для каждой границы
    remainder = total_tests % len(borders)    # Остаток тестов
    return [
        base_count + (1 if i < remainder else 0) for i in range(len(borders))
    ]
```

File: src/prog_questions/QuestionRandomCondition.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+class QuestionRandomCondition(QuestionBase):
+
+    def __init__(self, *, seed: int, condition_length: int, array_length: int, strictness: float):
+        super().__init__(seed=seed, condition_length=condition_length, array_length=array_length, strictness=strictness)
+        self.task = Task(array_length, condition_length, seed)
+        self.sample_code = self.task.code
+        self.parse(self.task.text)
+        self.seed = seed
+
```
Comment (NaGIBIN-IGOR) [inline src/prog_questions/QuestionRandomCondition.py:81]: Добавить настраиваемый параметр space_amount

File: src/prog_questions/QuestionRandomCondition.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+            else:
+                return f"Ошибка: ожидалось '{expected_output}', получено '{output}'"
+        except CompilationError as e:
+            return f"Ошибка компиляции: {e}"
+        except ExecutionError as e:
+            return f"Ошибка выполнения (код {e.exit_code}): {e}"
+
+    # form test: INT edge case
+    def test_int_edge_case(self, code: str) -> str:
+        upper_edge = self.test_case([10 ** 12] * self.parameters['array_length'], code, 1)
```
Comment (NaGIBIN-IGOR) [inline src/prog_questions/QuestionRandomCondition.py:142]: Не забыть заменить хардкод 1 на self.space_amount

**Review by NaGIBIN-IGOR — CHANGES_REQUESTED** [review]

File: src/prog_questions/QuestionRandomCondition.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    <table border="1">
+        <thead>
+            <tr>
+                <th>Входные данные</th>
+                <th>Выходные данные</th>
+            </tr>
+        </thead>
+        <tbody>
+            <tr>
+                <td>[35, 7, 93, 66, 62, 48, 53, 5, 18, 21]</td>
```
[the remark this answers is not in the document]
Comment (dechlorine) [reply src/prog_questions/QuestionRandomCondition.py]: Убрал запятые и скобки. Теперь студент будет видеть только числа через пробел.

File: src/prog_questions/QuestionRandomCondition.py (hunk 2 above)
Comment (dechlorine) [reply src/prog_questions/QuestionRandomCondition.py]: Добавил комментарий.

File: src/prog_questions/QuestionRandomCondition.py (hunk 3 above)
Comment (dechlorine) [reply src/prog_questions/QuestionRandomCondition.py:147]: Я решил сделать 16 тестов на краевые случаи, из них:
8 - с одинаковым числом (10^2, 10^4, 10^7 и 10^12), в частности последний проверяет также выход за INT
8 - с чередованием больших и маленьких чисел.

Остальные тесты будут рандомизированные, в том числе покрывающие устойчивость к нескольким пробелам (если установлена достаточная въедливость)

Comment (dechlorine) [reply src/prog_questions/QuestionRandomCondition.py:148]: Тесты добавлены.

File: src/prog_questions/QuestionRandomCondition.py (hunk 6 above)
Comment (dechlorine) [reply src/prog_questions/QuestionRandomCondition.py:149]: Исправлено.

File: src/prog_questions/QuestionRandomCondition.py (hunk 4 above)
Comment (dechlorine) [reply src/prog_questions/QuestionRandomCondition.py]: Граничных тестов всегда одинаковое число (потому что они обязательные и входят в минимальное тестирование). В отдельный метод вынесу распределение рандомных тестов.

File: src/prog_questions/QuestionRandomCondition.py (hunk 5 above)
Comment (dechlorine) [reply src/prog_questions/QuestionRandomCondition.py:87]: Зависимость количества пробелов от въедливости реализована в тестировании программы.

**Review by NaGIBIN-IGOR — APPROVED** [review]
